### ISSUE TITLE
Supply useful arguments to the Flux.train! callback function `cb`.

### ISSUE THREAD

**Issue by roberthoenig** [FluxML/Flux.jl#489]
`Flux.train!` takes an optional argument `cb`. `cb` is a function that gets called after every training pass. 

Currently, `Flux.train!` doesn't supply any useful arguments to `cb`. This means that it can be used for simple callbacks like `cb = throttle(() -> println("training"), 10)`. However, `cb` cannot evaluate any interesting information from the evaluation of the last training pass, like the model input and output, the loss, etc.

It'd be great if Flux.train! could (perhaps on demand?) supply `cb` with useful info about the last training pass. The question now is what can be considered "useful". 

Here are two parameters I'd find useful:

* the current iteration count
* the loss for the last training pass

**Comment by MikeInnes** [comment]
I think the way this will look is that `cb` becomes a function of one argument. You can easily ignore it (`_ -> println("training")`) but if you don't you'll get some context object, and you can do `cx -> @show cx.epochs`, `cx.loss` etc.

Suggestions for what else it should contain are welcome. cc @avik-pal who needed something like this.

**Comment by MikeInnes** [comment]
@roberthoenig would you be up for putting together a PR?

**Comment by roberthoenig** [comment]
Yup, can do over the weekend.

**Comment by dhairyagandhi96** [comment]
I was thinking if people would like to see some micro benchmarks built in - time taken to train the last epoch comes to mind.

**Comment by ToucheSir** [comment]
Ref. https://github.com/FluxML/Flux.jl/pull/498#issuecomment-860934819

**Comment by CarloLucibello** [comment]
FastAI.jl provides more high level functionality

**Comment by DhairyaLGandhi** [comment]
Again, that doesn't address the fact that we should do better.

**Comment by CarloLucibello** [comment]
Closing as resolved. `Flux.train!` no longer takes a callback at all: passing `cb` raises an error pointing to a manual `gradient`/`update!` loop, which is the recommended way to get at per-step information.

For the specific things asked for here, `Flux.trainstep!` returns the loss for the step, and a loss function may return `(loss, aux...)` so any intermediate quantities you want are handed back to you directly.
